lazy_dataset/core.py

```diff
@@ -1922,77 +1996,48 @@ def indexable(self):
             found_bucket = False
-            for j, (
-                    bucket, creation_idx, lower_bound, upper_bound, max_value
-            ) in enumerate(buckets):
-                if lower_bound <= value <= upper_bound:
-                    bucket.append(example)
-                    max_value = max(max_value, value)
-                    if (
-                            len(bucket) >= self.batch_size
-                            or (
-                            self.total_size_threshold is not None
-                            and (len(
-                        bucket) * max_value) > self.total_size_threshold
-                    )
-                    ):
-                        if self.sort_by_key:
-                            bucket = sorted(bucket, key=self.key, reverse=True)
-                        yield bucket
+            for j, (bucket, _) in enumerate(buckets):
+                found_bucket = bucket.maybe_append(example)
+                if found_bucket:
+                    if bucket.is_completed():
+                        data = bucket.data
+                        if self.sort_key is not None:
+                            data = sorted(data, key=self.sort_key, reverse=self.reverse_sort)
+                        yield data
                         buckets.pop(j)
-                    else:
-                        lower_bound = max(
-                            lower_bound, value * (1 - self.max_padding_rate)
-                        )
-                        upper_bound = min(
-                            upper_bound, value / (1 - self.max_padding_rate)
-                        )
-                        buckets[j] = (
-                            bucket, creation_idx,
-                            lower_bound, upper_bound, max_value
-                        )
-                    found_bucket = True
                     break
             if not found_bucket:
-                buckets.append((
-                    [example],
-                    i,
-                    value * (1 - self.max_padding_rate),
-                    value / (1 - self.max_padding_rate),
-                    value
-                ))
+                new_bucket = self.bucket_cls(
+                    example, *self.bucket_args, **self.bucket_kwargs
+                )
+                buckets.append((new_bucket, i))
 
             if self.expiration is not None:
-                expired = set()
-                for j, (bucket, creation_idx, *_) in enumerate(buckets):
-                    if (i - creation_idx) / self.batch_size >= self.expiration:
-                        # do not count batches to not get into a blocking state
-                        # if no batch gets ever completed. However also do not
-                        # operate with example counts to prevent the expiration
-                        # parameter being batch_size dependent
-                        expired.add(j)
+                for j in list(range(len(buckets)))[::-1]:
```

Comment:
Is it necessary to iterate in reverse order through the buckets?
You can at most drop one example, so a break after the pop would be fine